warehouse/oidc/services.py

```diff
@@ -287,11 +310,30 @@ def find_publisher(
             publisher = find_publisher_by_issuer(
                 self.db, self.issuer_url, signed_claims, pending=pending
             )
+
+            jwt_token_identifier: str | None = signed_claims.get("jti")
+            # jti is in the __preverified_claims__ set, so if it was present,
+            # it was already checked
+            if pending is False and jwt_token_identifier:
+                if self.token_identifier_exists(jwt_token_identifier):
+                    self.metrics.increment(
+                        "warehouse.oidc.reused_token",
+                        tags=metrics_tags,
+                    )
+                    raise ReusedTokenError("JWT Token already used to mint a token.")
```

Comment:
I was kind of expecting this logic to live within a verifier for the `jti` claim, and for claim verification to just fail if the identifier had been seen before. Is there a reason we're doing this here instead? It feels a bit out of place to be part of the publisher lookup.

Reply:
@DarkaMaul can offer his rationale as well, but my thinking has two parts:

1. Consistency/single sourcing: not every supported OIDC IdP uses the `jti` claim, so keeping the check at the service/lookup level avoids us needing to add the check explicitly to each publisher provider. We could probably handle that equivalently with a mixin to minimize duplication, however.
2. Claim verification is currently independent of the current `request` object, so we'd have to do a larger refactor to plumb `request` (or the appropriate service) into `check_jti` or similar to handle this on the per-claim level.

Reply:
The first one seems to be a case for adding per-provider `jti` claim verification instead :)

Not sure I follow the second one, what is the `request` object needed for? The configuration for the redis connection?

Reply:
> The first one seems to be a case for adding per-provider `jti` claim verification instead :)

True!

> Not sure I follow the second one, what is the request object needed for? The configuration for the redis connection?

Yep -- I think we need it in some form, either to create an appropriate `IOIDCPublisherService` _or_ to directly connect to Redis (in which case we need the `request` to access the appropriate Redis URL via `request.registry.settings["oidc.jwk_cache_url"]`.

Now that I think about it, I guess we don't actually need that though -- we can pass `self` (i.e. the service) as a new parameter to `publisher.verify_claims`, and claim verification can use that service to check the JTI.

Reply:
Implemented in [b405fae](https://github.com/pypi/warehouse/pull/16254/commits/b405faed27746d8e6de41b7e8651a266e0023ab2)